Instead of directly uploading the package downloaded added the option…

geloro94 opened the pull request:
… to persistate.

And loading a package from a persistence directory.

geloro94 committed:
Instead of directly uploading the package downloaded added the option to persistate.
And loading a package from a persistence directory.

jpwiedekopf commented on README.md:
```diff
@@ -45,6 +45,7 @@ usage: fhir_populator [-h] --endpoint ENDPOINT [--authorization-header AUTHORIZA
                       [--log-level {INFO,WARNING,DEBUG,ERROR}] [--rewrite-versions] [--only-put] [--versioned-ids]
                       [--exclude-resource-type [EXCLUDE_RESOURCE_TYPE ...] | --only [ONLY ...]]
                       [--registry-url REGISTRY_URL] [--package PACKAGES [PACKAGES ...]]
+                      [--persistate] [--from-persistence] [--persistence-dir PERSISTANCE_DIR]
```
it's `persist`, not `persistate`. Plz change everywhere!

jpwiedekopf commented on src/fhir_populator/populator.py:
```diff
@@ -263,7 +269,7 @@ def parse_args() -> argparse.Namespace:
         )
         parser.add_argument(
-            "--endpoint", required=True, type=str,
+            "--endpoint", type=str, default="",
```
`default = None`

jpwiedekopf commented on src/fhir_populator/populator.py:
```diff
@@ -424,9 +442,79 @@ def get_latest_package_version(self, package_name: str) -> str:
+            """
+            Recursively creates directories from src to dst
+            """
+            for root, dirs, files in os.walk(src):
```
nested for loops are not necessary, `os.makedirs` is already recursive (like `mkdir -p`)

jpwiedekopf commented on src/fhir_populator/populator.py:
```diff
@@ -424,9 +442,79 @@ def get_latest_package_version(self, package_name: str) -> str:
+                for file in files:
+                    src_file = os.path.join(root, file)
+                    dst_file = os.path.join(dst, os.path.relpath(src_file, src))
+                    shutil.copy2(src_file, dst_file)
```
[`shutil.copytree`](https://docs.python.org/3/library/shutil.html#shutil.copytree) can copy an entire tree (and create intermediate dirs as needed, maing a call to `os.makedirs` redundant as well.

jpwiedekopf commented on src/fhir_populator/populator.py:
```diff
@@ -424,9 +442,79 @@ def get_latest_package_version(self, package_name: str) -> str:
     def populate(self) -> None:
-        packages = self.resolve_package_versions()
-        dependency_graph = self.download_packages(packages)
+        packages = None
```
do you see a use case where you restore from persistence, but don't want everything? If you use the same persistence dir for multiple calls (for whatever reason) you currently would restore all packages. Currently, the `package` argument is ignored for the call to `self.load_persisted`, I don't think that's ideal.

jpwiedekopf reviewed (changes requested):
good functionality, but some changes are needed still.

geloro94 committed:
Fixed some minor issues.
Effectively now has 2 modes:
From packages hosting service (simplifier) by default,
which can be persisted.
And from persisted.

When packages are persisted they are all added to one dependency_graph.json. If non exists its created. Else they are composed.

When packages are loaded from persistence the latest local version is obtained if no specific version is defined.
Afterwards the package information and if specified its dependencies are extracted from the dependency_graph.json.

jpwiedekopf reviewed (approved):
:+1:
